vllm/v1/worker/gpu_model_runner.py

```diff
@@ -341,12 +351,21 @@ def _update_states(self, scheduler_output: "SchedulerOutput") -> bool:
 
             # Update the persistent batch.
             self.input_batch.num_computed_tokens_cpu[req_index] = (
-                req_data.num_computed_tokens)
+                num_computed_tokens)
             start_index = len(req_state.block_ids) - len(
                 req_data.new_block_ids)
             self.input_batch.block_table.append_row(req_index, start_index,
                                                     req_data.new_block_ids)
-
+            if new_token_ids:
+                start_token_index = num_computed_tokens
+                end_token_index = num_computed_tokens + num_new_tokens
+                self.input_batch.token_ids_cpu[
+                    req_index,
+                    start_token_index:end_token_index] = new_token_ids
+                self.input_batch.num_tokens[req_index] += num_new_tokens
+
+        # Check if the batch has changed. If not, we can skip copying the
+        # sampling metadata from CPU to GPU.
         batch_changed = len(removed_req_indices) > 0 or len(req_ids_to_add) > 0
```

Comment:
Just to make sure. Is this still correct? Even we have the same batch of requests, we may still need to copy the new tokens?

Reply:
To clarify, this comment means that we don't need to copy the "sampling metadata" (e.g., temperature, top_p, etc.) which is constant during the request's entire lifetime.